### PR TITLE
Feature - Input Action Refactor

### PR DESCRIPTION
**Pull request by StephenHodgson**
# XRTK - Mixed Reality Toolkit Change Request

## Overview

Feature Change Request for Input Actions Refactoring.

## Target of the change:

Is this enhancement for:

- Core (core framework, interfaces and definitions)
- Extension

## Changes:

Brief list of the targeted features that are being changed.

- Fixes: #111
- Fixes: #289 
- Fixes: #209 

## Changes

- Added new `AxisConstraintAttribute` to help better define `MixedRealityInputActions` in the inspectors.
- Added a new `MixedRealityInputActionDropdown` class to help draw the input action dropdown in custom inspectors.

## Breaking Changes:

Are there any breaking changes included in this change that would prevent or cause issue for existing projects.

- Removed `MixedRealityInputActionRulesProfile`
- Removed `IInputActionRule`
- Removed `InputActionRuleDigital`
- Removed `InputActionRuleDualAxis`
- Removed `InputActionRulePoseAxis`
- Removed `InputActionRuleQuaternionAxis`
- Removed `InputActionRuleSingleAxis`
- Removed `InputActionRuleVectorAxis`
- Removed `MixedRealityInputSystemProfile.InputActionRulesProfile`
- Removed `MixedRealityInspectorUtility.CheckMixedRealityConfigured`

### CODE REVIEW

File: XRTK-Core/Packages/com.xrtk.core/Definitions/InputSystem/MixedRealityInputAction.cs [hunk 1]
```diff
@@ -30,32 +48,61 @@ public MixedRealityInputAction(uint id, string description, AxisType axisConstra
         /// <summary>
         /// Default input action that doesn't represent any defined action.
         /// </summary>
-        public static MixedRealityInputAction None { get; } = new MixedRealityInputAction(0, "None");
+        public static MixedRealityInputAction None { get; } = new MixedRealityInputAction(default, 0, "None");
```
Comment (Jarodshow) [inline XRTK-Core/Packages/com.xrtk.core/Definitions/InputSystem/MixedRealityInputAction.cs:51]: Could this be `readonly` so it be can be compile time constant?

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Definitions/InputSystem/MixedRealityInputAction.cs:51]: That's a c#8 language feature

Comment (Jarodshow) [inline XRTK-Core/Packages/com.xrtk.core/Definitions/InputSystem/MixedRealityInputAction.cs:51]:
```suggestion
        public static readonly MixedRealityInputAction None = new MixedRealityInputAction(default, 0, "None");
```

Comment (Jarodshow) [reply XRTK-Core/Packages/com.xrtk.core/Definitions/InputSystem/MixedRealityInputAction.cs:51]: Sorry, I meant without the getter

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Definitions/InputSystem/MixedRealityInputAction.cs:51]: This will not compile until Unity supports c# 8

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Definitions/InputSystem/MixedRealityInputAction.cs:51]: Oh, I see, this is now a field.

**Comment by StephenHodgson** [comment]
@SimonDarksideJ do you want me to split the rules removal from this PR to speed up the approval process for the other import changes?

**Comment by SimonDarksideJ** [comment]
@StephenHodgson can you elaborate a bit more on the changes and the difference in usage for this PR.  A mini-guide for how you see input actions being consumed in the platform.

And can we add some more Unit Tests to validate input actions assignment and use, including programmatic configuration and comparison?

**Comment by SimonDarksideJ** [comment]
Additionally, can you check current implementations, such as the manipulation handler for the axis constraint, as after the change this still lists all actions for all properties.  Just as an example

**Comment by StephenHodgson** [comment]
I didn't fundamentally change how they work to warrant a whole new guide on their usage, lol.

Yes, I'll add a test for the backwards compatibility of the `MixedRealityInputAction.Equals()` of an action who doesn't have guids set vs one that does.

As for the property drawers and attributes example usages, I did add those to each of the class' documentation headers.

**Comment by SimonDarksideJ** [comment]
The recent change seemed to have broken something
1. Clear existing config
2. Create new scene
3. COnfigure Scene
4. Add GO with manipulation handler
5. Observe Inspector

image


Also error
```
ArgumentException: id cannot be 0
XRTK.Definitions.InputSystem.MixedRealityInputAction..ctor (System.Guid profileGuid, System.UInt32 id, System.String description, XRTK.Definitions.Utilities.AxisType axisConstraint) (at Packages/com.xrtk.core/Definitions/InputSystem/MixedRealityInputAction.cs:62)
XRTK.Inspectors.PropertyDrawers.MixedRealityInputActionDropdown.OnGui (UnityEngine.Rect rect, UnityEditor.SerializedProperty property, UnityEngine.GUIContent content, XRTK.Definitions.Utilities.AxisType axisConstraintFilter) (at Packages/com.xrtk.core/Inspectors/PropertyDrawers/MixedRealityInputActionDropdown.cs:81)
XRTK.Inspectors.PropertyDrawers.InputActionPropertyDrawer.OnGUI (UnityEngine.Rect rect, UnityEditor.SerializedProperty property, UnityEngine.GUIContent content) (at Packages/com.xrtk.core/Inspectors/PropertyDrawers/InputActionPropertyDrawer.cs:17)
UnityEditor.PropertyDrawer.OnGUISafe (UnityEngine.Rect position, UnityEditor.SerializedProperty property, UnityEngine.GUIContent label) (at C:/buildslave/unity/build/Editor/Mono/ScriptAttributeGUI/PropertyDrawer.cs:23)
UnityEditor.PropertyHandler.OnGUI (UnityEngine.Rect position, UnityEditor.SerializedProperty property, UnityEngine.GUIContent label, System.Boolean includeChildren, UnityEngine.Rect visibleArea) (at C:/buildslave/unity/build/Editor/Mono/ScriptAttributeGUI/PropertyHandler.cs:139)
UnityEditor.GenericInspector.OnOptimizedInspectorGUI (UnityEngine.Rect contentRect) (at C:/buildslave/unity/build/Editor/Mono/Inspector/GenericInspector.cs:102)
UnityEditor.UIElements.InspectorElement+<CreateIMGUIInspectorFromEditor>c__AnonStorey1.<>m__0 () (at C:/buildslave/unity/build/Editor/Mono/Inspector/InspectorElement.cs:456)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```

**Comment by StephenHodgson** [comment]
Ah those are places where I need to stop creating new input actions and use the `None` provided lol

**Comment by StephenHodgson** [comment]
@SimonDarksideJ fixed

**Pull request merged**